Feature/support macos builds

ClausKlein opened the pull request:
Edit from river:
Closes #22 .
Closes #64 .
Closes #65 .
Closes #67 .

wusatosi commented:
note: Should close: #22 after merging.

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -12,7 +12,7 @@ on:
 jobs:
   preset-test:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-latest # 22.04
```
Note: Ubuntu-latest will switch to 24.04 in Jan 2025, see: https://github.com/actions/runner-images/issues/10636

ClausKlein replied:
and next than you will use ubuntu-22.04?

wusatosi replied:
yes. I am trying to point out that this comment will become out of date soon. I do not foresee our pipeline to fail at 24.04.

ClausKlein replied:
OK

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -31,13 +31,14 @@ jobs:
-            c: gcc
-          - cpp: clang++
-            c: clang
-        cpp_version: [17, 20, 23, 26]
```
Exemplar officially support C++ 17 with the intention to explore the possibilities to support earlier versions. 

Please keep the lowest matrix version 17.

ClausKlein replied:
Wow will you develop code for C++26 if it should compile with c++17?

ClausKlein replied:
done

wusatosi commented on .gitignore:
Is this change intended? 

I don't think current CMake is configured to generate compile_commands.json.

ClausKlein replied:
but it should to prepare the use of `run-clang-tidy `

wusatosi reviewed (changes requested):
Thanks for the PR, it looks like this PR is still a work in progress.

Please make sure the lowest C++ version in CI is C++17.

Would also recommend you provide more context for this pr.

ClausKlein commented on .github/workflows/ci_tests.yml:
```diff
@@ -46,31 +47,31 @@ jobs:
+          cmakeVersion: "~3.30.0"
           ninjaVersion: "^1.11.1"
       - name: Print installed softwares
         run: |
```
done

wusatosi replied:
no I mean to update the `Print installed softwares` 
maybe update it to:
```
${{ matrix.compiler.c }}  --version
${{ matrix.compiler.cpp }} --version
cmake --version
ninja --version
```

ClausKlein replied:
done, as far as it is possible!

wusatosi reviewed (changes requested):
Again thank you so much for the PR, CI testing is painful on GitHub so thank you for keeping the commit log clean, we can also squash merge it when it's done if you don't want to keep squashing your commits (I presume that's what all the force pushes are for).

Suggested changes:
- remove the ubuntu 20.04 comment from Ubuntu-latest
- revert the minimum version back to 17
- leave some documentation for the commented-out code, or remove them entirely.

The changes to the CI looks good to me besides the suggestions. The changes to the CMake files looks like great improvements, but I am not qualified to review them. Code owners mentioned in the PR would be better candidates to evaluate those changes.

steve-downey commented on CMakeLists.txt:
```diff
@@ -9,9 +9,17 @@ project(
+    message(FATAL_ERROR "In-source builds are not allowed!")
+endif()
+
+include(CPack)
```
Are we using cpack for anything yet?

ClausKlein replied:
just to be prepared to create a binary distribution.

steve-downey commented on CMakeLists.txt:
```diff
@@ -9,9 +9,17 @@ project(
 
+add_subdirectory(src/beman/exemplar)
+
+option(BUILD_TESTING "build tests too" ${PROJECT_IS_TOP_LEVEL})
```
The recommended pattern for options is to include the name of the project, so that projects can compose. Defaulting to is top level is the right thing, but we wouldn't want to build the tests of beman_interface if we included that. 
Was going to come back to fix, but if we're touching this, might as well.

neatudarius replied:
@ClausKlein , please update this PR with latest main.

ClausKlein replied:
done

steve-downey commented on CMakePresets.json:
```diff
@@ -15,7 +15,7 @@
       "cacheVariables": {
         "CMAKE_BUILD_TYPE": "Debug",
-        "CMAKE_CXX_FLAGS": "-fsanitize=address -fsanitize=pointer-compare -fsanitize=pointer-subtract -fsanitize=leak -fsanitize=undefined"
+        "CMAKE_CXX_FLAGS": "-fsanitize=address -fsanitize=pointer-compare -fsanitize=pointer-subtract -fsanitize=undefined"
```
Why remove the leak sanitizer?

wusatosi replied:
I think this is a bundled fix for #65

ClausKlein replied:
`gcc` on **OSX** does not support this!

steve-downey replied:
Is this because there's a `g++` that isn't actually GCC? 
Least common denominator options is unfortunate.

steve-downey commented on src/beman/exemplar/CMakeLists.txt:
```diff
@@ -1,9 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-add_library(beman.exemplar STATIC)
+add_library(beman.exemplar INTERFACE)
```
If we change this we should enable VERIFY_INTERFACE_HEADER_SETS at the same time. Compiling the header is the purpose of having the cpp file. 

Also, header only is more of an accident than a plan, and since this also serves as a template for further work I'm not sure if INTERFACE should be the default?

ClausKlein replied:
see too https://github.com/beman-project/exemplar/issues/59

wusatosi commented:
I presume by comment: https://github.com/beman-project/exemplar/issues/64#issuecomment-2451745051

This PR packs a fix for #64 as well? 
Is this why we are using interface build?

ClausKlein commented on .github/workflows/ci_tests.yml:
```diff
@@ -46,24 +46,24 @@ jobs:
-            cmake_args:
-              description: "Dynamic"
-              cmake_args: "-DBUILD_SHARED_LIBS=on"
+          # - platform: ubuntu-22.04
+          #   compiler:
+          #     cpp: g++
+          #     c: gcc
+          #   cpp_version: 20
+          #   cmake_args:
+          #     description: "Dynamic"
+          #     cmake_args: "-DBUILD_SHARED_LIBS=on"
```
yes

wusatosi replied:
Ah I think you can simply remove this :0, I see the comment here, both works.

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -46,24 +46,24 @@ jobs:
+          #     cmake_args: "-DBUILD_SHARED_LIBS=on"
 
-    name: "Bulid & Test: ${{ matrix.compiler.c }} ${{ matrix.cpp_version }} ${{ matrix.cmake_args.description }}"
+    name: "Bulid & Test on ${{ matrix.platform }}: ${{ matrix.compiler.c }} ${{ matrix.cpp_version }} ${{ matrix.cmake_args.description }}"
```
Thanks for the update!

ClausKlein replied:
done

neatudarius replied:
I would prefer shorter this part `name: Continuous Integration Tests` -> `name: CI Tests`

wusatosi replied:
> I would prefer shorter this part `name: Continuous Integration Tests` -> `name: CI Tests`

Updating the name workflow name will cause a new Actions tab and could get cluttered up real quick (e.g. linter-checks).

![image](https://github.com/user-attachments/assets/f09ad1ee-0016-49c0-a735-1ee594979f26)

I think `Continuous Integration Tests` is good enough and could stay as-is. It is out-of-scope for this PR and there's a lot of bundled changes in this PR already.

wusatosi commented on CMakePresets.json:
```diff
@@ -56,7 +56,11 @@
       "name": "gcc-release",
-      "configurePreset": "gcc-release"
+      "configurePreset": "gcc-release",
+      "targets": [
+        "all_verify_interface_header_sets",
+        "all"
+      ]
```
Is there a reason why this is not applied in `gcc-debug`?

ClausKlein replied:
It is verified too while building tests and examples.

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -74,36 +75,46 @@ jobs:
         run: |
-          clang++ --version
-          g++ --version
+          ${{ matrix.compiler.cpp }} --version || ${CXX} --version || echo ignored
+          ${{ matrix.compiler.c }} --version || ${CC} --version || echo ignored
```
```suggestion
          ${{ matrix.compiler.cpp }} --version || echo ignored
          ${{ matrix.compiler.c }} --version || echo ignored
```

I don't think `CXX` and `CC` environment variable are not defined throughout the action.
It sucks that we have to do this for macos though.

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -46,24 +46,25 @@ jobs:
               cpp: g++
               c: gcc
-            cpp_version: 17
+            cpp_version: 20
```
```suggestion
            cpp_version: 17
```

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -46,24 +46,25 @@ jobs:
-            cmake_args:
-              description: "Dynamic"
-              cmake_args: "-DBUILD_SHARED_LIBS=on"
+          # NOTE: This example is an header only library!
+          # - platform: ubuntu-22.04
+          #   compiler:
+          #     cpp: g++
+          #     c: gcc
+          #   cpp_version: 20
+          #   cmake_args:
+          #     description: "Dynamic"
+          #     cmake_args: "-DBUILD_SHARED_LIBS=on"
```
```suggestion
```
Alternative to remove this section, I don't mind just leaving this in though.

neatudarius commented on .github/workflows/ci_tests.yml:
```diff
@@ -46,24 +46,25 @@ jobs:
+          #     cmake_args: "-DBUILD_SHARED_LIBS=on"
 
-    name: "Bulid & Test: ${{ matrix.compiler.c }} ${{ matrix.cpp_version }} ${{ matrix.cmake_args.description }}"
+    name: "Test on ${{ matrix.platform }}: ${{ matrix.compiler.c }} ${{ matrix.cpp_version }} ${{ matrix.cmake_args.description }}"
```
```suggestion
    name: "${{ matrix.platform }}: ${{ matrix.compiler.c }} ${{ matrix.cpp_version }} ${{ matrix.cmake_args.description }}"
```

Same here: It's enough to see `CI Tests / ubuntu-24.04: gcc-14 c++17 ..."

neatudarius reviewed (commented):
LGTM for current status, but I have some small observations

wusatosi commented on .gitignore:
```diff
@@ -1,2 +1,8 @@
+/CMakeFiles/
+/CMakeUserPresets.json
 /build
+/stagedir
```
What is this directory for?

ClausKlein replied:
to use for `DESTDIR=$PWD/stagedir ninja -C build install` instead to `/tmp`

ClausKlein replied:
see i.e.: https://github.com/beman-project/execution26/blob/main/cmake/CMakeGenericPresets.json

wusatosi replied:
Thanks for the context!

wusatosi reviewed (approved):
CI part looks good to me besides the suggestions (nitpicks) above.

Thanks for the contribution.

wusatosi commented on CMakeLists.txt:
```diff
@@ -26,14 +34,14 @@ if(BUILD_TESTING)
         set(INSTALL_GTEST OFF) # Disable GoogleTest installation
         set(BUILD_TESTING OFF) # Disable GoogleTest tests
+        set(BUILD_GMOCK OFF)
+        set(CMAKE_CXX_FLAGS) # FIXME: Do not use our project settings!
```
```suggestion
        set(CMAKE_CXX_FLAGS) # Prevent propagating our project settings
```
I don't think `FIXME` is appropriate here, feel free to reject this. This is a nitpick.

ClausKlein replied:
I do not agree with you:
```
bash-5.2$ cd build/gcc-debug/
bash-5.2$ ninja clean gtest
[1/1] Cleaning all built files...
Cleaning... 9 files.
bash-5.2$ ninja -nv gtest
[1/2] /usr/bin/g++  -I/Users/clausklein/Workspace/cpp/beman-project/exemplar/build/gcc-debug/_deps/googletest-src/googletest/include -I/Users/clausklein/Workspace/cpp/beman-project/exemplar/build/gcc-debug/_deps/googletest-src/googletest -fsanitize=address -fsanitize=pointer-compare -fsanitize=pointer-subtract -fsanitize=undefined -g -std=c++17 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.1.sdk -mmacosx-version-min=14.7  -DGTEST_HAS_PTHREAD=1 -MD -MT _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o -MF _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o.d -o _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o -c /Users/clausklein/Workspace/cpp/beman-project/exemplar/build/gcc-debug/_deps/googletest-src/googletest/src/gtest-all.cc
[2/2] : && /usr/local/bin/cmake -E rm -f lib/libgtest.a && /usr/bin/ar qc lib/libgtest.a  _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o && /usr/bin/ranlib lib/libgtest.a && /usr/local/bin/cmake -E touch lib/libgtest.a && :
bash-5.2$ 
```

wusatosi replied:
okay cool!

ClausKlein committed:
Add support to build on OSX with clang-18 or newer

This is an interface library (header only)

Prevent in source build

Fix macos CI build

Use all_verify_interface_header_sets on CI too.

Co-authored-by: Darius Neațu <neatudarius@users.noreply.github.com>

Co-authored-by: River <26424577+wusatosi@users.noreply.github.com>

ClausKlein committed:
Quickfix

ClausKlein committed:
Add more FIXME notes

ClausKlein commented on .github/workflows/ci_tests.yml:
```diff
@@ -31,39 +31,32 @@ jobs:
-            args: "-DCMAKE_CXX_FLAGS=-fsanitize=thread"
-          - description: "ASan"
-            args: "-DCMAKE_CXX_FLAGS='-fsanitize=address -fsanitize=undefined'"
+          # FIXME: find a better solution to be usable for multible OS builds on CI! CK
+          # - description: "TSan"
+          #  args: "-DCMAKE_CXX_FLAGS=-fsanitize=thread"
+          # - description: "ASan"
+          #   args: "-DCMAKE_CXX_FLAGS='-fsanitize=address -fsanitize=undefined'"
```
better solution  may be found here: https://github.com/beman-project/execution26/tree/main/.github/workflows

wusatosi replied:
Can you provide more description here? What are you trying to do?

wusatosi replied:
Hey CK, I would suggest you defer further improvement to another PR. Leaving a note here with some explanation on what improvements would be needed should be sufficient.

This PR is currently already complex and changes suggested to the build system are affecting fundamentals to our setup, it may be best to leave this PR in a stable state so other beman project members can perform reviews. As it may take some time for members to discuss the implications of the build updates.

Having small, atomic PRs also help all of us manage changes easily, as the longer this PR is hanging in the review queue, the more out of date this PR is in respect to other pending improvements, this applies to both sides. 

Again thank you for your contribution.

ClausKlein replied:
Yes, at the moment, I have no plan to change more on this **PR**.
But there are to **FIXME** in my changes and I have no advice how to continue?

For me, your **CI** workflow tries to match.

I would split the CI in **OS** specific parts, reduce the compiler versions, and the C++ standards you want to check.
Also I would also create **CI** specific `cmake workflow presets` and use them.

see too [CMakeWorkflowPresets: How to keep it as simple as possible?](https://discourse.cmake.org/t/cmakeworkflowpresets-how-to-keep-it-as-simple-as-possible/8735)

ClausKlein commented on CMakeLists.txt:
```diff
@@ -26,14 +34,14 @@ if(BUILD_TESTING)
         set(INSTALL_GTEST OFF) # Disable GoogleTest installation
         set(BUILD_TESTING OFF) # Disable GoogleTest tests
+        set(BUILD_GMOCK OFF)
+        set(CMAKE_CXX_FLAGS) # FIXME: Do not propagate our project settings! CK
```
What if your build dependencies doesn't compile with your compiler flags?

steve-downey replied:
Then the code is likely fundamentally incompatible? Except for warning flags, pretty much all other flags are detectable changes that will compile differently, leading to ODR violations, sometimes just harmless. This is true even for the standard library to some extent, although standard library implementors go to great lengths to avoid such breaks. 
Feature test macros and polyfills are common examples. Standard library implementation choice (libc++ vs libstdc++) is another. 
It's possible that some flags are peculiar to a package, or there are exceptional workarounds necessary, but by default all C++ in a link context should be compiled with the same flags.

ClausKlein replied:
see https://github.com/beman-project/exemplar/pull/66#discussion_r1835769955

neatudarius commented on .github/workflows/ci_tests.yml:
```diff
@@ -74,31 +67,41 @@ jobs:
-          cmake --install build --config Release --prefix /opt/beman.exemplar
-          find /opt/beman.exemplar -type f
+          cmake --install build --config Release --prefix /tmp/beman.exemplar
+          find /tmp/beman.exemplar -type f
```
Any reason not to keep /opt? It's usually a path used for Linux libraries and I think that path should also exist for MAC OS.

ClausKlein replied:
Yes, but `/opt` is owned by root.
I would use `${PWD}/stagedir` like boost does.

neatudarius replied:
I think we supposed regardless of who is owner of /opt, all users can create a subdirectory or whatever.

My point is that /opt is used in all repos in Beman right now in documentation. If this recommendation is not OK, I'm OK to change it everywhere (other PRs for docs), but if it's works, I won't do the change.

I will also test and come back.

ClausKlein replied:
### from CI build on MAC OS:
```bash
ld: warning: ignoring duplicate libraries: 'lib/Release/libgtest.a'
[1/1] Building CXX object src/beman/exemplar/CMakeFiles/beman.exemplar_verify_interface_header_sets.dir/Release/beman.exemplar_verify_interface_header_sets/beman/exemplar/identity.hpp.cxx.o
CMake Error at build/src/beman/exemplar/cmake_install.cmake:44 (file):
  file cannot create directory: /opt/beman.exemplar/include/beman/exemplar.
  Maybe need administrative privileges.
Call Stack (most recent call first):
  build/cmake_install.cmake:42 (include)


Error: Process completed with exit code 1.
```

neatudarius replied:
```bash
dariusn@pc:~/git/Beman/iterator_interface$ ls -l / | grep opt
drwxr-xr-x   7 dariusn root       4096 Nov 12 07:38 opt
```
I guess it depends if and how you have configured `/opt`.  So we may add:
```
# Create the /opt directory if it doesn’t already exist
sudo mkdir -p /opt

# Change ownership of /opt to the current user
sudo chown -R $USER /opt
```
(check other existing commands, sudo may not be required!)


I would keep the current PR just for adding MAC OS support. We can discuss on discourse (https://discourse.bemanproject.org/latest) in a new post if `/opt` is or not the best path to recommend to install in docs, etc CI.

We should aim to have the same default path in all places/repos! (docs, CI files, etc). It just makes the usability simpler for new users! And it still allows folks to use other custom install path on own system!

ClausKlein replied:
It does not successfully run on **Darwin** if we install to `/opt`?

As a user of project, I want to have control if and where the binaries are installed!

I am not willing to change the.

steve-downey replied:
/opt is the correct location in a unix FHS style system because every other prefix is owned by someone else. / and /usr (now fused anyway) are owned by the OS distro, /usr/local is the machine admin, ~ and ~/.local are user controlled.
https://www.pathname.com/fhs/pub/fhs-2.3.html#OPTADDONAPPLICATIONSOFTWAREPACKAGES

"A package to be installed in /opt must locate its static files in a separate /opt/<package> or /opt/<provider> directory tree, where <package> is a name that describes the software package and <provider> is the provider's LANANA registered name."

So a todo to register `beman` with LANANA, although it's actually unlikely real world installs go there. 



Testing install in CI is also not locking in an install location for any user. PREFIX is controlled by the installer. 

That all said, outside of a scratch VM like a GitHub action runs in, using a directory inside out root, but outside the source tree, like we do for build, is the friendly way to do it. 

Staging isn't the right name, because it's not going anywhere after that. Trying to do a package manager's job is a mistake

neatudarius replied:
> As a user of project, I want to have control if and where the binaries are installed!
> 
> I am not willing to change more.
@ClausKlein , I agree with you, as a user you can install them anywhere you want. As a developer inside the Beman project, you must use some requirements/recommendations/best practices used inside the Beman org. That's not questionable, but you are more than welcome to propose to change / upgrade/ improves these sets of standards or practices.


TLDR:
* This CI config is exclusively for internal development. It must follow our internal practices or rules.
* Similarly, examples for root README.md, are just examples and our recommendations.  Our choice, but not an obligation for users.
* Any user can install our libraries in a custom path, that's for sure an available option!

Steve also explained this in previous comment - "Testing install in CI is also not locking in an install location for any user. PREFIX is controlled by the installer."

I hope that is clear now. Please revert the changes related to the install path. Thanks!

ClausKlein committed:
Merge branch 'main' into feature/support-macos-builds

neatudarius reviewed (changes requested):
Thanks for contribution @ClausKlein!

I'm temporary blocking this PR until we finish pending discussions, mostly related to Beman org practices. Thanks for understanding!

ClausKlein committed:
Revert my workaround and notes

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -31,39 +31,31 @@ jobs:
+          args: "-DCMAKE_CXX_FLAGS=-fsanitize=thread"
           - description: "ASan"
-            args: "-DCMAKE_CXX_FLAGS='-fsanitize=address -fsanitize=undefined'"
+          args: "-DCMAKE_CXX_FLAGS='-fsanitize=address -fsanitize=undefined'"
```
```suggestion
            args: "-DCMAKE_CXX_FLAGS='-fsanitize=address -fsanitize=undefined'"
```

wusatosi commented on .github/workflows/ci_tests.yml:
```diff
@@ -31,39 +31,31 @@ jobs:
             args: ""
           - description: "TSan"
-            args: "-DCMAKE_CXX_FLAGS=-fsanitize=thread"
+          args: "-DCMAKE_CXX_FLAGS=-fsanitize=thread"
```
```suggestion
            args: "-DCMAKE_CXX_FLAGS=-fsanitize=thread"
```

ClausKlein commented:
@wusatosi I closed this PR because with only one CI workflow file, it seems to me to complicated to support **Windows**, **OSX**, and `Linux`.

First of all, the `cmake workflow presets` and maybe the `toolchain` files must be available.
Second, it must be clear how to create the `cmake export config package`.

wusatosi commented:
> @wusatosi I closed this PR because with only one CI workflow file, it seems to me to complicated to support **Windows**, **OSX**, and `Linux`.
> 
> First of all, the `cmake workflow presets` and maybe the `toolchain` files must be available. Second, it must be clear how to create the `cmake export config package`.

I don't get what are you trying to say here. 

You are free to create a separate CI file/ separate CI job under the same file. I don't have opinions with either approaches, I will go with whatever reviewers would prefer, #80 is still pending and I am fine to the change MSVC related change to another CI job/ another workflow.

My personal take is keeping all configuration in one file reduces the amount of workflow on the GitHub panel. It shows what configuration is run on which platform in one file so there's less chance that one set of matrix is missing from anther platform due to not checking across files. It also allows downstream to copy-paste that one file to launch another set of matrix easily, with clear examples on how to setup specific build commands on each platform.

Regarding first point:  What does this have to do with workflows? What toolchain files are you refereeing to.
Second point: What do you mean how to create the `cmake export config package`? are you referring to #31 ? I struggle to see relevance here, can you provide more description?

wusatosi commented:
Can I assume you are not going to take this PR forward? @ClausKlein
